modules/diskq/dqtool.c

```diff
@@ -241,12 +251,292 @@ _relocate_validate_options(void)
 }
 
+static gint
+_open_fd_for_reading(const gchar *fname)
+{
+  gint fd = open(fname, O_RDONLY);
+
+  if (fd == -1)
+    fprintf(stderr, "Failed to open %s, reason: %s\n", fname, strerror(errno));
+
+  return fd;
+}
+
+static gboolean
+_get_content_length_by_fd(gint fd, gsize *size)
+{
+  struct stat st;
+
+  if (fstat(fd, &st) == -1)
+    {
+      fprintf(stderr, "Error: %s\n", strerror(errno));
+      *size = 0;
+      return FALSE;
+    }
+
+  *size = st.st_size;
+  return TRUE;
+}
+
+static gboolean
+_mmap_content_from_fd(gint fd, gchar **content, gsize *content_length)
+{
+  if (!_get_content_length_by_fd(fd, content_length))
+    return FALSE;
+
+  *content = mmap(NULL, *content_length, PROT_READ, MAP_PRIVATE, fd, 0);
+  if (*content == MAP_FAILED)
+    {
+      fprintf(stderr, "mmap failed, reason: %s\n", strerror(errno));
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static gboolean
+_mmap_copy_to_fd(gint fd, gchar *content, gsize content_length)
+{
+  if (ftruncate(fd, content_length) == -1)
+    {
+      fprintf(stderr, "Error: %s\n", strerror(errno));
+      return FALSE;
+    }
+
+  gchar *dst_content = mmap(NULL, content_length, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
+  if (dst_content == MAP_FAILED)
+    {
+      fprintf(stderr, "Error: %s\n", strerror(errno));
+      return FALSE;
+    }
+
+  memcpy(dst_content, content, content_length);
+
+  if (msync(dst_content, content_length, MS_SYNC) == -1)
+    {
+      fprintf(stderr, "Error: %s\n", strerror(errno));
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static gint
+_open_fd_for_writing(const gchar *fname)
+{
+  gint fd = open(fname, O_RDWR | O_CREAT, S_IRUSR | S_IWUSR);
+
+  if (fd == -1)
+    fprintf(stderr, "Error: failed to create %s, reason: %s\n", fname, strerror(errno));
+
+  return fd;
+}
+
+static void
+_remove_file(const gchar *fname)
+{
+  if (remove(fname) == -1)
+    fprintf(stderr, "Failed to remove %s, reason; %s\n", fname, strerror(errno));
+}
+
+static gboolean
+_move_file_between_different_filesystems(const gchar *from, const gchar *to)
+{
+  gboolean result = TRUE;
+  int dst_fd = -1;
+
+  gint src_fd = _open_fd_for_reading(from);
+  if (src_fd == -1)
+    goto error;
+
+  gsize content_length = 0;
+  gchar *content = NULL;
+  if (!_mmap_content_from_fd(src_fd, &content, &content_length))
+    goto error;
+
+  dst_fd = _open_fd_for_writing(to);
+  if (dst_fd == -1)
+    goto error;
+
+  if (!_mmap_copy_to_fd(dst_fd, content, content_length))
+    goto error;
+
+  result = TRUE;
+  goto exit;
+error:
+  result = FALSE;
+exit:
+  if (src_fd != -1)
+    close(src_fd);
+  if (dst_fd != -1)
+    close(dst_fd);
+  if (result)
+    _remove_file(from);
+
+  return result;
+}
+
+static gboolean
+_move_file(const gchar *from, const gchar *to)
+{
+  gint rename_res = rename(from, to);
+
+  switch (rename_res)
+    {
+    case 0:
+      return TRUE;
+    case EXDEV:
```

Comment:
`EXDEV` is and errno value. `rename_res` is still -1 in that case. (switch-case on wrong variable).

Reply:
thx!